src/metabase/driver/mysql.clj

```diff
@@ -348,6 +353,40 @@
                                 2)
                           (hx/literal "-01"))))
 
+(defmethod sql.qp/->honeysql [:mysql :datetime-diff]
+  [driver [_ x y unit :as clause]]
+  (let [x (sql.qp/->honeysql driver x)
+        y (sql.qp/->honeysql driver y)
+        disallowed-types (keep
+                          (fn [v]
+                            (when-let [db-type (some-> v hx/type-info hx/type-info->db-type keyword str/upper-case)]
+                              (let [base-type (sql-jdbc.sync/database-type->base-type driver db-type)]
+                                (when-not (some #(isa? base-type %) [:type/Date :type/DateTime])
+                                  (name db-type)))))
+                          [x y])]
+    (when (seq disallowed-types)
+      (throw (ex-info (tru "Only datetime, timestamp, or date types allowed. Found {0}"
+                           (pr-str disallowed-types))
+                      {:found disallowed-types
+                       :type  qp.error-type/invalid-parameter})))
```

Comment:
This is the definition of `invalid-parameter`

```clj
(deferror invalid-parameter
  "The query is parameterized, and a supplied parameter has an invalid value."
  :parent invalid-query
  :show-in-embeds? true)
```

This means parameter in the Dashboard or SQL `{{template tag}}` parameter sense. It's not the right error type here. 

We're not really doing a lot with the error types at the moment -- the old BigQuery driver had logic to retry on certain query errors like timeouts but would not retry malformed queries -- don't remember if this is still in the new BigQuery driver or not. I don't think we're using them in too many other places. Either way we should still try to use error types that make sense so we can leave the door open to do something useful with that info in the future.


```suggestion
                      {:found disallowed-types
                       :type  qp.error-type/invalid-query})))
```

`invalid-query` is the closest thing we have here that makes sense. You could add a subtype called `type-mismatch` or something if you wanted. I'm cool either way. Just don't use `invalid-parameter` for something that doesn't have to do with query `:parameters`

Reply:
Makes sense. Updated in https://github.com/metabase/metabase/commit/6d0485c90dcc2aaff9b50448cbccc98f50f87b19